api/pkg/transformer/types/encoder/cyclical_encoder_test.go

```diff
@@ -0,0 +1,972 @@
+						Period: spec.PeriodType_WEEK,
+					},
+				},
+			},
+			expectedEncoder: &CyclicalEncoder{
+				Period: spec.PeriodType_UNDEFINED,
+				Min:    0,
+				Max:    WeekInSec,
+			},
+		},
+		{
+			desc: "Should succeed - by epoch time: MONTH",
+			config: &spec.CyclicalEncoderConfig{
+				EncodeBy: &spec.CyclicalEncoderConfig_ByEpochTime{
+					ByEpochTime: &spec.ByEpochTime{
+						Period: spec.PeriodType_MONTH,
+					},
+				},
+			},
+			expectedEncoder: &CyclicalEncoder{
+				Period: spec.PeriodType_MONTH,
+				Min:    0,
+				Max:    0,
+			},
+		},
+		{
+			desc: "Should succeed - by epoch time: QUARTER",
+			config: &spec.CyclicalEncoderConfig{
+				EncodeBy: &spec.CyclicalEncoderConfig_ByEpochTime{
+					ByEpochTime: &spec.ByEpochTime{
+						Period: spec.PeriodType_QUARTER,
+					},
+				},
+			},
+			expectedEncoder: &CyclicalEncoder{
+				Period: spec.PeriodType_QUARTER,
+				Min:    0,
+				Max:    0,
+			},
+		},
+		{
+			desc: "Should succeed - by epoch time: HALF",
+			config: &spec.CyclicalEncoderConfig{
+				EncodeBy: &spec.CyclicalEncoderConfig_ByEpochTime{
+					ByEpochTime: &spec.ByEpochTime{
+						Period: spec.PeriodType_HALF,
+					},
+				},
+			},
+			expectedEncoder: &CyclicalEncoder{
+				Period: spec.PeriodType_HALF,
+				Min:    0,
+				Max:    0,
+			},
+		},
+		{
+			desc: "Should succeed - by epoch time: YEAR",
+			config: &spec.CyclicalEncoderConfig{
+				EncodeBy: &spec.CyclicalEncoderConfig_ByEpochTime{
+					ByEpochTime: &spec.ByEpochTime{
+						Period: spec.PeriodType_YEAR,
+					},
+				},
+			},
+			expectedEncoder: &CyclicalEncoder{
+				Period: spec.PeriodType_YEAR,
+				Min:    0,
+				Max:    0,
+			},
+		},
+		{
+			desc: "Should fail - by epoch time: unsupported cycle period",
+			config: &spec.CyclicalEncoderConfig{
+				EncodeBy: &spec.CyclicalEncoderConfig_ByEpochTime{
+					ByEpochTime: &spec.ByEpochTime{
+						Period: 8,
+					},
+				},
+			},
+			expectedEncoder: nil,
+			expectedErr:     fmt.Errorf("invalid or unspported cycle period"),
+		},
+		{
+			desc: "Should fail - by epoch time: missing cycle period",
+			config: &spec.CyclicalEncoderConfig{
+				EncodeBy: &spec.CyclicalEncoderConfig_ByEpochTime{
+					ByEpochTime: &spec.ByEpochTime{},
+				},
+			},
+			expectedEncoder: nil,
+			expectedErr:     fmt.Errorf("invalid or unspported cycle period"),
+		},
+		{
+			desc: "Should fail - by epoch time: invalid config",
+			config: &spec.CyclicalEncoderConfig{
+				EncodeBy: &spec.CyclicalEncoderConfig_ByEpochTime{},
+			},
+			expectedEncoder: nil,
+			expectedErr:     fmt.Errorf("cyclical encoding config invalid or undefined"),
+		},
+	}
+	for _, tC := range testCases {
+		t.Run(tC.desc, func(t *testing.T) {
+			gotEncoder, err := NewCyclicalEncoder(tC.config)
+			if tC.expectedErr != nil {
+				assert.EqualError(t, err, tC.expectedErr.Error())
+			} else {
+				assert.Equal(t, tC.expectedEncoder, gotEncoder)
+			}
+		})
+	}
+}
+
+func TestCyclicalEncode(t *testing.T) {
```

Comment:
can you add test where each input in `reqValues` have at least other input that fall in the same cycle.